cloud_governance/policy/ibm/ibm_cost_report.py

```diff
@@ -0,0 +1,121 @@
+
+    @typechecked
+    def collect_tags_from_machines(self, tags: list):
+        """
+        This method return tags from list of string tags
+        @param tags:
+        @return:
+        """
+        hardware_tags = {}
+        if tags:
+            for tag in tags:
+                if ':' in tag:
+                    key, value = tag.split(':')
+                    if key in self.owned_tags:
+                        hardware_tags[key] = value
+        return hardware_tags
+
+    def get_hardware_data(self):
+        """
+        This method returns the baremetal resource data
+        @return:
+        """
+        bare_metals = self.classic_operations.get_hardware_ids()
+        collect_machines_data = {}
+        for hardware in bare_metals:
+            hardware_tags = self.collect_tags_from_machines(
+                tags=self.classic_operations.get_hardware_tags(hardware_id=hardware.get('id')))
+            hardware_tags['fqdn'] = hardware.get('fullyQualifiedDomainName').lower()
+            collect_machines_data[hardware_tags['fqdn']] = hardware_tags
+        return collect_machines_data
+
+    def get_virtual_machine_data(self):
+        """
+        This method returns the virtual machine data
+        @return:
+        """
+        vms = self.classic_operations.get_virtual_machine_ids()
+        collect_machines_data = {}
+        for vm in vms:
+            vm_tags = self.collect_tags_from_machines(
+                tags=self.classic_operations.get_virtual_machine_tags(vm_id=vm.get('id')))
+            vm_tags['fqdn'] = vm.get('fullyQualifiedDomainName').lower()
+            collect_machines_data[vm_tags['fqdn']] = vm_tags
+        return collect_machines_data
+
+    @typechecked
+    def get_invoice_data(self, collect_machines_data: dict):
+        """
+        This method returns the invoice data
+        @return:
+        """
+        month, year = self.month, self.year
+        if not month and not year:
+            date = datetime.datetime.now().date()
+            month, year = date.month, date.year
+        else:
+            month = int(month)
+            year = int(year)
+        invoice_data, users = self.ibm_account.get_invoice_data(month=month, year=year)
+        for fqdn in invoice_data.keys():
+            if fqdn not in collect_machines_data:
+                user = [user for user in users if user in fqdn]
+                if user:
+                    user_tags = self.ibm_account.get_user_tags_from_gsheet(username=f'{user[0]}@redhat.com', user_email='yes')
+                    invoice_data[fqdn].update(self.collect_tags_from_machines(user_tags))
+        return invoice_data
+
+    @typechecked
+    def beautify_and_aggregate_data(self, resource_data: dict, invoice_data: dict):
+        """
+        This method concatenate two dicts and aggregate its sum
+        @param resource_data:
+        @param invoice_data:
+        @return:
+        """
+        pd1 = pandas.DataFrame(list(resource_data.values()))
```

Comment:
You should use meaningful name and not pd1, pd2
maybe invoice_df, resource_df

Reply:
ok